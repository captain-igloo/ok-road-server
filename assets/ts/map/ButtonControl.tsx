import { IconProp } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Control, ControlOptions, DomUtil } from 'leaflet';
import * as React from 'react';
import { Button } from 'react-bootstrap';
import { createPortal } from 'react-dom';

interface ButtonControlOptions extends ControlOptions {
    className?: string;
    icon: IconProp;
    onClick: () => void;
    title?: string;
    variant?: string;
}

export default class ButtonControl extends Control {
    private portal: React.ReactNode;

    private onClick: () => void;

    private icon: IconProp;

    private className?: string;

    private div?: HTMLDivElement;

    private title?: string;

    private variant?: string;

    public constructor(options: ButtonControlOptions) {
        super(options);
        this.onClick = options.onClick;
        this.icon = options.icon;
        this.className = options.className;
        this.title = options.title;
        this.variant = options.variant;
    }

    private getDiv(): HTMLDivElement {
        if (!this.div) {
            this.div = DomUtil.create('div');
        }
        return this.div;
    }

    private createPortal() {
        this.portal = createPortal(
            <Button
                onClick={this.onClick}
                title={this.title}
                variant={this.variant || 'light'}
            >
                <FontAwesomeIcon className={this.className} icon={this.icon} />
            </Button>,
            this.getDiv(),
        );
    }

    public setClassName(className?: string) {
        this.className = className;
        this.createPortal();
    }

    public setVariant(variant?: string) {
        this.variant = variant;
        this.createPortal();
    }

    public setOnClick(onClick: () => void) {
        this.onClick = onClick;
        this.createPortal();
    }

    public setTitle(title?: string) {
        this.title = title;
        this.createPortal();
    }

    public setIcon(icon: IconProp) {
        this.icon = icon;
        this.createPortal();
    }

    public onAdd() {
        this.createPortal();
        return this.getDiv();
    }

    public getPortal() {
        return this.portal;
    }
}
