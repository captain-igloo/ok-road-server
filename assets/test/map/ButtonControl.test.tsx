import '../setup';

import { faArrowRotateRight } from '@fortawesome/free-solid-svg-icons/faArrowRotateRight';
import { faDrawPolygon } from '@fortawesome/free-solid-svg-icons/faDrawPolygon';
import {
    describe,
    expect,
    jest,
    test,
} from '@jest/globals';
import { fireEvent, render } from '@testing-library/react';

import ButtonControl from '../../ts/map/ButtonControl';

describe('ButtonControl', () => {
    test('ButtonControl should render properly', () => {
        const buttonControl = new ButtonControl({
            className: 'className',
            icon: faDrawPolygon,
            onClick: () => {},
            title: 'title',
            variant: 'variant',
        });
        buttonControl.addTo({
            _controlCorners: {
                topright: {
                    appendChild: () => {},
                },
            },
            on: () => {},
        });
        render(buttonControl.getPortal());
        expect((buttonControl as any).div).toMatchSnapshot();
    });

    test('setClassName() should set className', () => {
        const buttonControl = new ButtonControl({
            icon: faDrawPolygon,
            onClick: () => {},
        });
        buttonControl.setClassName('className');
        render(buttonControl.getPortal());
        expect((buttonControl as any).div).toMatchSnapshot();
    });

    test('setVariant() should set variant', () => {
        const buttonControl = new ButtonControl({
            icon: faDrawPolygon,
            onClick: () => {},
        });
        buttonControl.setVariant('variant');
        render(buttonControl.getPortal());
        expect((buttonControl as any).div).toMatchSnapshot();
    });

    test('setTitle() should set title', () => {
        const buttonControl = new ButtonControl({
            icon: faDrawPolygon,
            onClick: () => {},
        });
        buttonControl.setTitle('title');
        render(buttonControl.getPortal());
        expect((buttonControl as any).div).toMatchSnapshot();
    });

    test('setIcon() should set icon', () => {
        const buttonControl = new ButtonControl({
            icon: faDrawPolygon,
            onClick: () => {},
        });
        buttonControl.setIcon(faArrowRotateRight);
        render(buttonControl.getPortal());
        expect((buttonControl as any).div).toMatchSnapshot();
    });

    test('setOnClick() should set onClick', () => {
        const buttonControl = new ButtonControl({
            icon: faDrawPolygon,
            onClick: () => {},
            title: 'title',
        });
        const onClick = jest.fn();
        buttonControl.setOnClick(onClick);
        render(buttonControl.getPortal());
        fireEvent.click((buttonControl as any).div.getElementsByTagName('button').item(0));
        expect(onClick).toHaveBeenCalled();
    });
});
