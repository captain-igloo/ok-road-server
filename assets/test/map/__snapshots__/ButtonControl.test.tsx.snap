// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ButtonControl ButtonControl should render properly 1`] = `
<div
  class="leaflet-control"
>
  <button
    class="btn btn-variant"
    title="title"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-draw-polygon className"
      data-icon="draw-polygon"
      data-prefix="fas"
      role="img"
      viewBox="0 0 512 512"
    >
      <path
        d="M64 0c23.7 0 44.4 12.9 55.4 32l273.1 0c11.1-19.1 31.7-32 55.4-32 35.3 0 64 28.7 64 64 0 34-26.5 61.8-60 63.9l-2.1 2.9-71.5 99.1c3.6 8 5.6 16.9 5.6 26.2s-2 18.2-5.6 26.2l2.2 3.1 69.3 96 2.1 2.9c33.5 2.1 60 29.9 60 63.9 0 35.3-28.7 64-64 64-23.7 0-44.4-12.9-55.4-32l-273.1 0c-11.1 19.1-31.7 32-55.4 32-35.3 0-64-28.7-64-64 0-23.7 12.9-44.4 32-55.4l0-273.1C12.9 108.4 0 87.7 0 64 0 28.7 28.7 0 64 0zM394.2 413.4l-65.5-90.6-2.2-3.1c-2.1 .2-4.3 .3-6.5 .3-35.3 0-64-28.7-64-64s28.7-64 64-64c2.2 0 4.4 .1 6.5 .3l67.7-93.7c-.6-.9-1.1-1.7-1.6-2.6L119.4 96c-5.6 9.7-13.7 17.8-23.4 23.4l0 273.1c9.7 5.6 17.8 13.7 23.4 23.4l273.1 0c.5-.9 1.1-1.8 1.6-2.6z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`ButtonControl setClassName() should set className 1`] = `
<div
  class=""
>
  <button
    class="btn btn-light"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-draw-polygon className"
      data-icon="draw-polygon"
      data-prefix="fas"
      role="img"
      viewBox="0 0 512 512"
    >
      <path
        d="M64 0c23.7 0 44.4 12.9 55.4 32l273.1 0c11.1-19.1 31.7-32 55.4-32 35.3 0 64 28.7 64 64 0 34-26.5 61.8-60 63.9l-2.1 2.9-71.5 99.1c3.6 8 5.6 16.9 5.6 26.2s-2 18.2-5.6 26.2l2.2 3.1 69.3 96 2.1 2.9c33.5 2.1 60 29.9 60 63.9 0 35.3-28.7 64-64 64-23.7 0-44.4-12.9-55.4-32l-273.1 0c-11.1 19.1-31.7 32-55.4 32-35.3 0-64-28.7-64-64 0-23.7 12.9-44.4 32-55.4l0-273.1C12.9 108.4 0 87.7 0 64 0 28.7 28.7 0 64 0zM394.2 413.4l-65.5-90.6-2.2-3.1c-2.1 .2-4.3 .3-6.5 .3-35.3 0-64-28.7-64-64s28.7-64 64-64c2.2 0 4.4 .1 6.5 .3l67.7-93.7c-.6-.9-1.1-1.7-1.6-2.6L119.4 96c-5.6 9.7-13.7 17.8-23.4 23.4l0 273.1c9.7 5.6 17.8 13.7 23.4 23.4l273.1 0c.5-.9 1.1-1.8 1.6-2.6z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`ButtonControl setIcon() should set icon 1`] = `
<div
  class=""
>
  <button
    class="btn btn-light"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-arrow-rotate-right"
      data-icon="arrow-rotate-right"
      data-prefix="fas"
      role="img"
      viewBox="0 0 512 512"
    >
      <path
        d="M436.7 74.7L448 85.4 448 32c0-17.7 14.3-32 32-32s32 14.3 32 32l0 128c0 17.7-14.3 32-32 32l-128 0c-17.7 0-32-14.3-32-32s14.3-32 32-32l47.9 0-7.6-7.2c-.2-.2-.4-.4-.6-.6-75-75-196.5-75-271.5 0s-75 196.5 0 271.5 196.5 75 271.5 0c8.2-8.2 15.5-16.9 21.9-26.1 10.1-14.5 30.1-18 44.6-7.9s18 30.1 7.9 44.6c-8.5 12.2-18.2 23.8-29.1 34.7-100 100-262.1 100-362 0S-25 175 75 75c99.9-99.9 261.7-100 361.7-.3z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`ButtonControl setTitle() should set title 1`] = `
<div
  class=""
>
  <button
    class="btn btn-light"
    title="title"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-draw-polygon"
      data-icon="draw-polygon"
      data-prefix="fas"
      role="img"
      viewBox="0 0 512 512"
    >
      <path
        d="M64 0c23.7 0 44.4 12.9 55.4 32l273.1 0c11.1-19.1 31.7-32 55.4-32 35.3 0 64 28.7 64 64 0 34-26.5 61.8-60 63.9l-2.1 2.9-71.5 99.1c3.6 8 5.6 16.9 5.6 26.2s-2 18.2-5.6 26.2l2.2 3.1 69.3 96 2.1 2.9c33.5 2.1 60 29.9 60 63.9 0 35.3-28.7 64-64 64-23.7 0-44.4-12.9-55.4-32l-273.1 0c-11.1 19.1-31.7 32-55.4 32-35.3 0-64-28.7-64-64 0-23.7 12.9-44.4 32-55.4l0-273.1C12.9 108.4 0 87.7 0 64 0 28.7 28.7 0 64 0zM394.2 413.4l-65.5-90.6-2.2-3.1c-2.1 .2-4.3 .3-6.5 .3-35.3 0-64-28.7-64-64s28.7-64 64-64c2.2 0 4.4 .1 6.5 .3l67.7-93.7c-.6-.9-1.1-1.7-1.6-2.6L119.4 96c-5.6 9.7-13.7 17.8-23.4 23.4l0 273.1c9.7 5.6 17.8 13.7 23.4 23.4l273.1 0c.5-.9 1.1-1.8 1.6-2.6z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;

exports[`ButtonControl setVariant() should set variant 1`] = `
<div
  class=""
>
  <button
    class="btn btn-variant"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-draw-polygon"
      data-icon="draw-polygon"
      data-prefix="fas"
      role="img"
      viewBox="0 0 512 512"
    >
      <path
        d="M64 0c23.7 0 44.4 12.9 55.4 32l273.1 0c11.1-19.1 31.7-32 55.4-32 35.3 0 64 28.7 64 64 0 34-26.5 61.8-60 63.9l-2.1 2.9-71.5 99.1c3.6 8 5.6 16.9 5.6 26.2s-2 18.2-5.6 26.2l2.2 3.1 69.3 96 2.1 2.9c33.5 2.1 60 29.9 60 63.9 0 35.3-28.7 64-64 64-23.7 0-44.4-12.9-55.4-32l-273.1 0c-11.1 19.1-31.7 32-55.4 32-35.3 0-64-28.7-64-64 0-23.7 12.9-44.4 32-55.4l0-273.1C12.9 108.4 0 87.7 0 64 0 28.7 28.7 0 64 0zM394.2 413.4l-65.5-90.6-2.2-3.1c-2.1 .2-4.3 .3-6.5 .3-35.3 0-64-28.7-64-64s28.7-64 64-64c2.2 0 4.4 .1 6.5 .3l67.7-93.7c-.6-.9-1.1-1.7-1.6-2.6L119.4 96c-5.6 9.7-13.7 17.8-23.4 23.4l0 273.1c9.7 5.6 17.8 13.7 23.4 23.4l273.1 0c.5-.9 1.1-1.8 1.6-2.6z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;
